cpp/include/cudf/aggregation/host_udf.hpp

```diff
@@ -43,49 +43,141 @@ namespace CUDF_EXPORT cudf {
  */
 
 /**
- * @brief The interface for host-based UDF implementation.
+ * @brief The fundamental interface for host-based UDF implementation.
  *
- * An implementation of host-based UDF needs to be derived from this base class, defining
- * its own version of the required functions. In particular:
- *  - The derived class is required to implement `get_empty_output`, `operator()`, `is_equal`,
- *    and `clone` functions.
- *  - If necessary, the derived class can also override `do_hash` to compute hashing for its
- *    instance, and `get_required_data` to selectively access to the input data as well as
- *    intermediate data provided by libcudf.
+ * This class declares the functions `do_hash`, `is_equal`, and `clone` that must be defined in
+ * the users' UDF implementation. These function are required for libcudf aggregation framework
+ * to perform its operations.
+ */
+class host_udf_base {
+  // Declare constructor private to prevent the users from deriving from this class.
+ private:
+  host_udf_base() = default;  ///< Default constructor
+
+  // Only allow deriving from the structs below.
+  friend struct host_udf_reduction_base;
+  friend struct host_udf_segmented_reduction_base;
+  friend struct host_udf_groupby_base;
```

Comment:
Why have this restriction? In the admittedly unlikely case where we have some new udf aggregation base type, they're just going to come in here and add the `friend` declaration anyway.

Reply:
This restriction is to prevent the users from deriving from `host_cudf_base`. They must derive from one of these 3 base structs above.